3.3.9 build fail , theme-default version detect error?

Report:

Try convert from gitbook to githon.
https://github.com/tsuyoshicho/WhenScarletIceMelted/tree/feature/githon

Setup and `npm install` success for `githon 3.3.9` and others.

But, when `npm run build`(githon build) phase, error occur.

Platform: Windows, Linux
Node: v14.4.0

<details>
<summary>`Githon doesn't satisfy the requirements of this plugin: >=4.0.0-alpha`</summary>

```
WhenScarletIceMelted on  feature/githon via ⬢ v14.4.0 took 15s
⏰[ 2020-06-20T07:40:09 ] ❯ npm install

> node-sass@4.14.1 install D:\Data\ReposWork\git\WhenScarletIceMelted\node_modules\node-sass
> node scripts/install.js

Cached binary found at C:\Users\tsuyoshi\AppData\Roaming\npm-cache\node-sass\4.14.1\win32-x64-83_binding.node

> node-sass@4.14.1 postinstall D:\Data\ReposWork\git\WhenScarletIceMelted\node_modules\node-sass
> node scripts/build.js

Binary found at D:\Data\ReposWork\git\WhenScarletIceMelted\node_modules\node-sass\vendor\win32-x64-83\binding.node
Testing binary
Binary is fine
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@2.1.3 (node_modules\fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@2.1.3: wanted {"os":"darwin","arch":"any"} (current: {"os":"win32","arch":"x64"})

added 788 packages from 768 contributors and audited 790 packages in 378.876s

60 packages are looking for funding
  run `npm fund` for details

found 3 vulnerabilities (1 low, 2 high)
  run `npm audit fix` to fix them, or `npm audit` for details
WhenScarletIceMelted on  feature/githon via ⬢ v14.4.0 took 6m20s
⏰[ 2020-06-20T07:46:34 ] ❯ npm run build

> WhenScarletIceMelted@1.0.0 build D:\Data\ReposWork\git\WhenScarletIceMelted
> githon build

info: 9 plugins are installed
info: 9 explicitly listed
info: plugin "creativecommons is loaded
info: plugin "styles-sass-fix is loaded
info: plugin "puml is loaded
info: plugin "highlight is loaded
info: plugin "search is loaded
info: plugin "lunr is loaded
info: plugin "fontsettings is loaded
info: plugin "theme-default is loaded
Githon doesn't satisfy the requirements of this plugin: >=4.0.0-alpha
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! WhenScarletIceMelted@1.0.0 build: `githon build`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the WhenScarletIceMelted@1.0.0 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\tsuyoshi\AppData\Roaming\npm-cache\_logs\2020-06-19T22_46_47_721Z-debug.log
exit status 1
WhenScarletIceMelted on  feature/githon via ⬢ v14.4.0 took 2s
⏰[ 2020-06-20T07:46:47 ] ❯
```
</details>

--- a/packages/githon/lib/githon.js
+++ b/packages/githon/lib/githon.js
@@ -2,21 +2,26 @@ const semver = require("semver");
 const pkg = require("../package.json");
 
 const VERSION = pkg.version;
-const VERSION_STABLE = VERSION.replace(/\-(\S+)/g, "");
+const VERSION_STABLE = VERSION.replace(/-(\S+)/g, "");
 
 const START_TIME = new Date();
 
 /**
-    Verify that this gitbook version satisfies a requirement
-    We can't directly use samver.satisfies since it will break all plugins when gitbook version is a prerelease (beta, alpha)
+ Verify that this gitbook version satisfies a requirement
+ We can't directly use samver.satisfies since it will break all plugins when gitbook version is a prerelease (beta, alpha)
 
-    @param {String} condition
-    @return {Boolean}
-*/
+ @param {String} condition
+ @return {Boolean}
+ */
 function satisfies(condition) {
     // Test with real version
     if (semver.satisfies(VERSION, condition)) return true;
 
+    // If plugin require -alpha -beta engine, skip it
+    // https://github.com/githon/githon/issues/42
+    if (/-(\S+)$/.test(condition)) {
+        return true;
+    }
     // Test with future stable release
     return semver.satisfies(VERSION_STABLE, condition);
 }

--- a/packages/githon/lib/plugins/__tests__/validatePlugin.js
+++ b/packages/githon/lib/plugins/__tests__/validatePlugin.js
@@ -1,3 +1,4 @@
+const Immutable = require("immutable");
 const Promise = require("../../utils/promise");
 const Plugin = require("../../models/plugin");
 const validatePlugin = require("../validatePlugin");
@@ -14,5 +15,85 @@ describe("validatePlugin", () => {
                 return Promise();
             }
         );
+    });
+    test("alpha plugin can loaded plugin", () => {
+        const testPkgRequireAlphaVersion = {
+            title: "Disqus",
+            name: "gitbook-plugin-disqus",
+            description: "Disqus comments thread in the footer of your pages",
+            icon: "./icon.png",
+            main: "index.js",
+            browser: "./_assets/plugin.js",
+            version: "1.0.1",
+            engines: {
+                gitbook: ">=4.0.0-alpha",
+            },
+            dependencies: {
+                react: "^15.3.2",
+                "react-disqus-thread": "^0.4.0",
+            },
+            devDependencies: {
+                "gitbook-plugin": "^4.0.0-alpha.0",
+                eslint: "3.7.1",
+                "eslint-config-gitbook": "1.4.0",
+            },
+            homepage: "https://github.com/GitbookIO/plugin-disqus",
+            repository: {
+                type: "git",
+                url: "https://github.com/GitbookIO/plugin-disqus.git",
+            },
+            license: "Apache-2.0",
+            bugs: {
+                url: "https://github.com/GitbookIO/plugin-disqus/issues",
+            },
+            keywords: ["gitbook:social"],
+            scripts: {
+                "build-js": "gitbook-plugin build ./src/index.js ./_assets/plugin.js",
+                prepublish: "npm run build-js",
+            },
+            gitbook: {
+                properties: {
+                    shortName: {
+                        type: "string",
+                        title: "Website Shortname",
+                        description: "Unique identifier for your website as registered on Disqus",
+                        required: true,
+                    },
+                    useIdentifier: {
+                        type: "boolean",
+                        title: "Pass page identifier option to Disqus",
+                        default: false,
+                    },
+                },
+                page: {
+                    title: "Disqus configuration for this page",
+                    properties: {
+                        disqus: {
+                            type: "object",
+                            properties: {
+                                enabled: {
+                                    type: "boolean",
+                                    title: "Enable Disqus comments for this page.",
+                                    default: true,
+                                },
+                                identifier: {
+                                    type: "string",
+                                    title: "Identifier for this page.",
+                                    description: "Tells the Disqus service how to identify the current page",
+                                },
+                            },
+                        },
+                    },
+                },
+            },
+        };
+        const plugin = new Plugin({
+            name: "test",
+            version: "4.0.0",
+            path: "/",
+            package: Immutable.fromJS(testPkgRequireAlphaVersion),
+            content: Immutable.fromJS({}),
+        });
+        return validatePlugin(plugin);
     });
 });

--- a/packages/githon/lib/plugins/validatePlugin.js
+++ b/packages/githon/lib/plugins/validatePlugin.js
@@ -3,11 +3,11 @@ const githon = require("../githon");
 const Promise = require("../utils/promise");
 
 /**
-    Validate a plugin
+ Validate a plugin
 
-    @param {Plugin} plugin
-    @return {Promise<Plugin>}
-*/
+ @param {Plugin} plugin
+ @return {Promise<Plugin>}
+ */
 function validatePlugin(plugin) {
     const packageInfos = plugin.getPackage();
 
@@ -18,13 +18,22 @@ function validatePlugin(plugin) {
         packageInfos.get("engines") &&
         (packageInfos.get("engines").get("gitbook") || packageInfos.get("engines").get("githon"));
 
+    const pluginName = packageInfos.get("name") || "unknown plugin";
     if (!isValid) {
-        return Promise.reject(new Error(`Error loading plugin "${plugin.getName()}" at "${plugin.getPath()}"`));
+        return Promise.reject(new Error(`Error loading plugin "${pluginName}" at "${plugin.getPath()}"`));
     }
-
-    const engine = packageInfos.get("engines").get("gitbook");
-    if (!githon.satisfies(engine)) {
-        return Promise.reject(new Error(`Githon doesn't satisfy the requirements of this plugin: ${engine}`));
+    const gitbookVersion = packageInfos.get("engines").get("gitbook");
+    const githonVersion = packageInfos.get("engines").get("githon");
+    // support "gitbook" and "githon"
+    if (gitbookVersion && !githon.satisfies(gitbookVersion)) {
+        return Promise.reject(
+            new Error(`GitHon doesn't satisfy the requirements of this plugin: ${pluginName} require ${githonVersion}`)
+        );
+    }
+    if (githonVersion && !githon.satisfies(githonVersion)) {
+        return Promise.reject(
+            new Error(`GitHon doesn't satisfy the requirements of this plugin: ${pluginName} require ${githonVersion}`)
+        );
     }
 
     return Promise(plugin);

